File: ios/Shared/KeyboardPreferences.swift
import Foundation

/// Shared preferences manager for keyboard configuration
/// Uses App Groups to share data between main app and keyboard extension
class KeyboardPreferences {
    
    // IMPORTANT: This must match the App Group ID configured in Xcode
    // Format: group.<bundle-identifier>
    static let appGroupIdentifier = //"group.com.issieshapiro.issieboard"
    "group.org.issieshapiro.test"
    
    // Keys for storing preferences
    struct Keys {
        static let currentProfile = "currentProfile"
        static let keyboardConfig = "keyboardConfig"
        static let lastUpdateTime = "lastUpdateTime"
        static let selectedLanguage = "selectedLanguage"
        static let keyboardDimensions = "keyboardDimensions"
    }
    
    // Notification names for observing changes
    static let preferencesDidChangeNotification = NSNotification.Name("KeyboardPreferencesDidChange")
    static let keyboardDimensionsDidChangeNotification = NSNotification.Name("KeyboardDimensionsDidChange")
    
    private let userDefaults: UserDefaults?
    
    init() {
        // Use App Group UserDefaults for sharing between app and extension
        self.userDefaults = UserDefaults(suiteName: KeyboardPreferences.appGroupIdentifier)
        
        if userDefaults == nil {
            print("⚠️ Failed to initialize App Group UserDefaults. Make sure App Groups capability is enabled.")
        }
    }
    
    // MARK: - Profile Management
    
    var currentProfile: String? {
        get {
            return userDefaults?.string(forKey: Keys.currentProfile)
        }
        set {
            userDefaults?.set(newValue, forKey: Keys.currentProfile)
            userDefaults?.set(Date().timeIntervalSince1970, forKey: Keys.lastUpdateTime)
            userDefaults?.synchronize()
            notifyPreferencesChanged()
        }
    }
    
    // MARK: - Language Selection
    
    var selectedLanguage: String? {
        get {
            return userDefaults?.string(forKey: Keys.selectedLanguage)
        }
        set {
            userDefaults?.set(newValue, forKey: Keys.selectedLanguage)
            userDefaults?.set(Date().timeIntervalSince1970, forKey: Keys.lastUpdateTime)
            userDefaults?.synchronize()
            notifyPreferencesChanged()
        }
    }
    
    // MARK: - Last Update Time
    
    var lastUpdateTime: TimeInterval {
        return userDefaults?.double(forKey: Keys.lastUpdateTime) ?? 0
    }
    
    // MARK: - JSON Configuration Storage
    
    /// Store keyboard configuration as JSON string
    func setKeyboardConfigJSON(_ jsonString: String) {
        userDefaults?.set(jsonString, forKey: Keys.keyboardConfig)
        userDefaults?.set(Date().timeIntervalSince1970, forKey: Keys.lastUpdateTime)
        userDefaults?.synchronize()
        notifyPreferencesChanged()
    }
    
    /// Retrieve keyboard configuration as JSON string
    func getKeyboardConfigJSON() -> String? {
        return userDefaults?.string(forKey: Keys.keyboardConfig)
    }
    
    // MARK: - Profile Storage
    
    /// Store profile configuration as JSON string
    func setProfileJSON(_ jsonString: String, forKey key: String) {
        userDefaults?.set(jsonString, forKey: "profile_\(key)")
        userDefaults?.synchronize()
    }
    
    /// Retrieve profile configuration as JSON string
    func getProfileJSON(forKey key: String) -> String? {
        return userDefaults?.string(forKey: "profile_\(key)")
    }
    
    // MARK: - Generic String Storage

    /// Store a string value for a given key
    func setString(_ value: String, forKey key: String) {
        userDefaults?.set(value, forKey: key)
        userDefaults?.synchronize()
    }

    /// Retrieve a string value for a given key
    func getString(forKey key: String) -> String? {
        return userDefaults?.string(forKey: key)
    }

    // MARK: - Keyboard Dimensions

    /// Store keyboard dimensions (width, height, device info)
    func setKeyboardDimensions(width: CGFloat, height: CGFloat, device: String, orientation: String, keysetId: String) {
        let dimensions: [String: Any] = [
            "width": Double(width),
            "height": Double(height),
            "device": device,
            "orientation": orientation,
            "keysetId": keysetId,
            "timestamp": Date().timeIntervalSince1970
        ]

        if let jsonData = try? JSONSerialization.data(withJSONObject: dimensions, options: []),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            print("📐 [KeyboardPreferences] Saving dimensions JSON: \(jsonString)")
            userDefaults?.set(jsonString, forKey: Keys.keyboardDimensions)
            let syncSuccess = userDefaults?.synchronize() ?? false
            print("📐 [KeyboardPreferences] UserDefaults sync success: \(syncSuccess)")

            // Post Darwin notification for cross-process communication
            // This works between app and keyboard extension
            print("📐 [KeyboardPreferences] Posting Darwin notification...")
            CFNotificationCenterPostNotification(
                CFNotificationCenterGetDarwinNotifyCenter(),
                CFNotificationName("com.issieboard.dimensionsChanged" as CFString),
                nil,
                nil,
                true
            )
            print("📐 [KeyboardPreferences] Darwin notification posted")

            // Also post regular notification for in-process listeners
            DispatchQueue.main.async {
                NotificationCenter.default.post(
                    name: KeyboardPreferences.keyboardDimensionsDidChangeNotification,
                    object: nil,
                    userInfo: dimensions
                )
                print("📐 [KeyboardPreferences] Regular notification posted")
            }
        }
    }

    /// Retrieve keyboard dimensions as JSON string
    func getKeyboardDimensionsJSON() -> String? {
        return userDefaults?.string(forKey: Keys.keyboardDimensions)
    }

    // MARK: - Notification
    
    private func notifyPreferencesChanged() {
        // Post notification on main thread
        DispatchQueue.main.async {
            NotificationCenter.default.post(
                name: KeyboardPreferences.preferencesDidChangeNotification,
                object: nil,
                userInfo: [
                    "timestamp": Date().timeIntervalSince1970,
                    "profile": self.currentProfile ?? "",
                    "language": self.selectedLanguage ?? ""
                ]
            )
        }
    }
    
    // MARK: - Debugging
    
    func printAllPreferences() {
        print("📱 Keyboard Preferences:")
        print("  App Group: \(KeyboardPreferences.appGroupIdentifier)")
        print("  Current Profile: \(currentProfile ?? "none")")
        print("  Selected Language: \(selectedLanguage ?? "none")")
        print("  Last Update: \(Date(timeIntervalSince1970: lastUpdateTime))")
        print("  Has Config: \(getKeyboardConfigJSON() != nil)")
    }
    
    // MARK: - Clear Preferences
    
    func clearAll() {
        // Remove all keys from the App Group UserDefaults
        guard let defaults = userDefaults else { return }
        
        // Get all keys and remove them
        let dictionary = defaults.dictionaryRepresentation()
        for key in dictionary.keys {
            defaults.removeObject(forKey: key)
        }
        
        defaults.synchronize()
        notifyPreferencesChanged()
        
        print("🗑️ Cleared all \(dictionary.keys.count) preference keys")
    }
}

// MARK: - Preference Observer

/// Observer class for monitoring preference changes in keyboard extension
class KeyboardPreferenceObserver {
    
    private var lastKnownUpdateTime: TimeInterval = 0
    private var checkTimer: Timer?
    private let preferences: KeyboardPreferences
    private let onChange: () -> Void
    
    init(preferences: KeyboardPreferences, onChange: @escaping () -> Void) {
        self.preferences = preferences
        self.onChange = onChange
        self.lastKnownUpdateTime = preferences.lastUpdateTime
    }
    
    /// Start polling for changes (keyboard extensions can't use Darwin notifications)
    func startObserving(interval: TimeInterval = 0.5) {
        stopObserving()
        
        checkTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            self?.checkForChanges()
        }
    }
    
    /// Stop polling for changes
    func stopObserving() {
        checkTimer?.invalidate()
        checkTimer = nil
    }
    
    private func checkForChanges() {
        let currentUpdateTime = preferences.lastUpdateTime
        
        if currentUpdateTime > lastKnownUpdateTime {
            lastKnownUpdateTime = currentUpdateTime
            
            print("🔄 Keyboard preferences changed at \(Date(timeIntervalSince1970: currentUpdateTime))")
            preferences.printAllPreferences()
            
            onChange()
        }
    }
    
    deinit {
        stopObserving()
    }
}

File: android/app/src/main/java/com/issieboardng/shared/KeyboardPreferences.kt
package com.issieboardng.shared

import android.content.Context
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper

/**
 * Shared preferences manager for keyboard configuration
 * Port of ios/Shared/KeyboardPreferences.swift
 * 
 * Uses SharedPreferences with MODE_MULTI_PROCESS to share data between main app and keyboard service
 */
class KeyboardPreferences(context: Context) {
    
    companion object {
        // SharedPreferences name for sharing between app and keyboard
        const val PREFS_NAME = "issieboard_keyboard_prefs"
        
        // Keys for storing preferences
        object Keys {
            const val CURRENT_PROFILE = "currentProfile"
            const val KEYBOARD_CONFIG = "keyboardConfig"
            const val LAST_UPDATE_TIME = "lastUpdateTime"
            const val SELECTED_LANGUAGE = "selectedLanguage"
        }
    }
    
    @Suppress("DEPRECATION")
    private val sharedPrefs: SharedPreferences = context.getSharedPreferences(
        PREFS_NAME, 
        Context.MODE_MULTI_PROCESS  // Allow sharing between processes
    )
    
    // MARK: - Profile Management
    
    var currentProfile: String?
        get() = sharedPrefs.getString(Keys.CURRENT_PROFILE, null)
        set(value) {
            sharedPrefs.edit()
                .putString(Keys.CURRENT_PROFILE, value)
                .putLong(Keys.LAST_UPDATE_TIME, System.currentTimeMillis())
                .apply()
        }
    
    // MARK: - Language Selection
    
    var selectedLanguage: String?
        get() = sharedPrefs.getString(Keys.SELECTED_LANGUAGE, null)
        set(value) {
            sharedPrefs.edit()
                .putString(Keys.SELECTED_LANGUAGE, value)
                .putLong(Keys.LAST_UPDATE_TIME, System.currentTimeMillis())
                .apply()
        }
    
    // MARK: - Last Update Time
    
    val lastUpdateTime: Long
        get() {
            // Handle both String and Long formats (JS saves as String, native saves as Long)
            return try {
                sharedPrefs.getLong(Keys.LAST_UPDATE_TIME, 0)
            } catch (e: ClassCastException) {
                // If it's stored as String (from JS), parse it
                try {
                    sharedPrefs.getString(Keys.LAST_UPDATE_TIME, "0")?.toLongOrNull() ?: 0
                } catch (e2: Exception) {
                    0
                }
            } catch (e: Exception) {
                0
            }
        }
    
    // MARK: - JSON Configuration Storage
    
    /** Store keyboard configuration as JSON string */
    fun setKeyboardConfigJSON(jsonString: String) {
        sharedPrefs.edit()
            .putString(Keys.KEYBOARD_CONFIG, jsonString)
            .putLong(Keys.LAST_UPDATE_TIME, System.currentTimeMillis())
            .apply()
    }
    
    /** Retrieve keyboard configuration as JSON string */
    fun getKeyboardConfigJSON(): String? {
        return sharedPrefs.getString(Keys.KEYBOARD_CONFIG, null)
    }
    
    // MARK: - Profile Storage
    
    /** Store profile configuration as JSON string */
    fun setProfileJSON(jsonString: String, key: String) {
        sharedPrefs.edit()
            .putString("profile_$key", jsonString)
            .apply()
    }
    
    /** Retrieve profile configuration as JSON string */
    fun getProfileJSON(key: String): String? {
        return sharedPrefs.getString("profile_$key", null)
    }
    
    // MARK: - Generic String Storage
    
    /** Store a string value for a given key */
    fun setString(value: String, key: String) {
        sharedPrefs.edit()
            .putString(key, value)
            .apply()
    }
    
    /** Retrieve a string value for a given key */
    fun getString(key: String): String? {
        return sharedPrefs.getString(key, null)
    }
    
    // MARK: - Debugging
    
    fun printAllPreferences() {
        alwaysLog("📱 Keyboard Preferences:")
        alwaysLog("  Prefs Name: $PREFS_NAME")
        alwaysLog("  Current Profile: ${currentProfile ?: "none"}")
        alwaysLog("  Selected Language: ${selectedLanguage ?: "none"}")
        alwaysLog("  Last Update: ${java.util.Date(lastUpdateTime)}")
        alwaysLog("  Has Config: ${getKeyboardConfigJSON() != null}")
    }
    
    // MARK: - Clear Preferences
    
    fun clearAll() {
        val keyCount = sharedPrefs.all.size
        sharedPrefs.edit().clear().apply()
        alwaysLog("🗑️ Cleared all $keyCount preference keys")
    }
}

/**
 * Observer class for monitoring preference changes in keyboard service
 * Port of iOS KeyboardPreferenceObserver
 */
class KeyboardPreferenceObserver(
    private val preferences: KeyboardPreferences,
    private val onChange: () -> Unit
) {
    
    private var lastKnownUpdateTime: Long = 0
    private var handler: Handler? = null
    private var checkRunnable: Runnable? = null
    
    init {
        lastKnownUpdateTime = preferences.lastUpdateTime
    }
    
    /** Start polling for changes (keyboard services can't easily use ContentObserver across processes) */
    fun startObserving(intervalMs: Long = 500) {
        stopObserving()
        
        handler = Handler(Looper.getMainLooper())
        checkRunnable = object : Runnable {
            override fun run() {
                checkForChanges()
                handler?.postDelayed(this, intervalMs)
            }
        }
        handler?.postDelayed(checkRunnable!!, intervalMs)
    }
    
    /** Stop polling for changes */
    fun stopObserving() {
        checkRunnable?.let { handler?.removeCallbacks(it) }
        checkRunnable = null
        handler = null
    }
    
    private fun checkForChanges() {
        val currentUpdateTime = preferences.lastUpdateTime
        
        if (currentUpdateTime > lastKnownUpdateTime) {
            lastKnownUpdateTime = currentUpdateTime
            
            alwaysLog("🔄 Keyboard preferences changed at ${java.util.Date(currentUpdateTime)}")
            preferences.printAllPreferences()
            
            onChange()
        }
    }
}